File: crates/haze-playout/src/bridge.rs
use std::collections::HashMap;
use std::path::PathBuf;
use std::sync::Arc;
use std::time::Duration;

use anyhow::{Context, Result};
use serde::Deserialize;
use serde_json::{json, Value};
use tokio::io::{AsyncBufReadExt, AsyncWriteExt, BufReader};
use tokio::net::tcp::{OwnedReadHalf, OwnedWriteHalf};
use tokio::net::TcpStream;
use tokio::sync::{mpsc, oneshot, Mutex};
use tokio::time::{sleep, timeout};

type BridgeResult<T> = std::result::Result<T, String>;
type PendingReply<T> = oneshot::Sender<BridgeResult<T>>;
type PendingMap<T> = Arc<Mutex<HashMap<String, PendingReply<T>>>>;
type PendingSynth = PendingMap<String>;
type PendingProducts = PendingMap<RenderedProduct>;

const SYNTH_REPLY_TIMEOUT: Duration = Duration::from_secs(90);
const PRODUCT_REPLY_TIMEOUT: Duration = Duration::from_secs(10);

#[derive(Clone)]
pub(crate) struct BridgeClient {
    writer: Arc<Mutex<OwnedWriteHalf>>,
    pending_synth: PendingSynth,
    pending_products: PendingProducts,
}

pub(crate) struct BridgeConnection {
    pub(crate) client: BridgeClient,
    pub(crate) events: mpsc::Receiver<Value>,
}

#[derive(Debug, Clone)]
pub(crate) struct SynthJob {
    pub(crate) id: String,
    pub(crate) text: String,
    pub(crate) reader_id: String,
    pub(crate) language: String,
    pub(crate) output_path: PathBuf,
}

#[derive(Debug, Clone)]
pub(crate) struct ProductRenderRequest {
    pub(crate) request_id: String,
    pub(crate) feed_id: String,
    pub(crate) package_id: String,
    pub(crate) force: bool,
}

#[allow(dead_code)]
#[derive(Debug, Clone, Default, Deserialize)]
pub(crate) struct RenderedProduct {
    #[serde(default)]
    pub(crate) id: String,
    #[serde(default)]
    pub(crate) feed_id: String,
    #[serde(default)]
    pub(crate) package_id: String,
    #[serde(default)]
    pub(crate) title: String,
    #[serde(default)]
    pub(crate) text: String,
    #[serde(default)]
    pub(crate) reader_id: String,
    #[serde(default)]
    pub(crate) language: String,
    #[serde(default)]
    pub(crate) metadata: HashMap<String, String>,
}

pub(crate) async fn connect(addr: &str) -> Result<BridgeConnection> {
    let stream = TcpStream::connect(addr)
        .await
        .with_context(|| format!("failed to connect to host bridge at {addr}"))?;
    let (reader, writer) = stream.into_split();
    let (tx, rx) = mpsc::channel(256);
    let pending_synth = Arc::new(Mutex::new(HashMap::new()));
    let pending_products = Arc::new(Mutex::new(HashMap::new()));
    tokio::spawn(read_loop(
        reader,
        tx,
        Arc::clone(&pending_synth),
        Arc::clone(&pending_products),
    ));
    Ok(BridgeConnection {
        client: BridgeClient {
            writer: Arc::new(Mutex::new(writer)),
            pending_synth,
            pending_products,
        },
        events: rx,
    })
}

pub(crate) async fn connect_retry(addr: &str) -> Result<BridgeConnection> {
    loop {
        match connect(addr).await {
            Ok(connection) => return Ok(connection),
            Err(err) => {
                tracing::warn!("waiting for host event bridge at {addr}: {err}");
                sleep(Duration::from_secs(1)).await;
            }
        }
    }
}

impl BridgeClient {
    pub(crate) async fn publish(&self, mut value: Value) -> Result<()> {
        if value.get("timestamp").is_none() {
            value["timestamp"] = json!(chrono::Utc::now().to_rfc3339());
        }
        let mut raw = serde_json::to_vec(&value)?;
        raw.push(b'\n');
        let mut writer = self.writer.lock().await;
        writer
            .write_all(&raw)
            .await
            .context("failed to write host event bridge event")
    }

    pub(crate) async fn service_ready(&self, feeds: usize) {
        let _ = self
            .publish(json!({
                "type": "service.ready",
                "source": "haze-playout",
                "data": {
                    "service": "haze-playout",
                    "feeds": feeds,
                }
            }))
            .await;
    }

    pub(crate) async fn synthesize(&self, job: SynthJob) -> Result<String> {
        let job_id = fallback_text(
            &job.id,
            &format!(
                "tts-{}",
                chrono::Utc::now().timestamp_nanos_opt().unwrap_or(0)
            ),
        );
        let (tx, rx) = oneshot::channel();
        self.pending_synth.lock().await.insert(job_id.clone(), tx);
        if let Some(parent) = job.output_path.parent() {
            tokio::fs::create_dir_all(parent)
                .await
                .with_context(|| format!("failed to create {}", parent.display()))?;
        }
        let publish_result = self
            .publish(json!({
                "type": "tts.synthesize",
                "source": "haze-playout",
                "subject": job_id,
                "data": {
                    "job_id": job_id,
                    "text": job.text,
                    "reader_id": job.reader_id,
                    "language": job.language,
                    "output_path": job.output_path,
                }
            }))
            .await;
        if let Err(err) = publish_result {
            self.pending_synth.lock().await.remove(&job_id);
            return Err(err);
        }
        let result = match timeout(SYNTH_REPLY_TIMEOUT, rx).await {
            Ok(Ok(result)) => result,
            Ok(Err(_)) => {
                self.pending_synth.lock().await.remove(&job_id);
                anyhow::bail!("event bridge closed while waiting for TTS");
            }
            Err(_) => {
                self.pending_synth.lock().await.remove(&job_id);
                anyhow::bail!("TTS request timed out");
            }
        };
        match result {
            Ok(path) => Ok(path),
            Err(err) => anyhow::bail!("{err}"),
        }
    }

    pub(crate) async fn render_product(
        &self,
        request: ProductRenderRequest,
    ) -> Result<RenderedProduct> {
        let request_id = fallback_text(
            &request.request_id,
            &format!(
                "product-{}-{}-{}",
                request.feed_id,
                request.package_id,
                chrono::Utc::now().timestamp_nanos_opt().unwrap_or(0)
            ),
        );
        let (tx, rx) = oneshot::channel();
        self.pending_products
            .lock()
            .await
            .insert(request_id.clone(), tx);
        let publish_result = self
            .publish(json!({
                "type": "product.render.request",
                "source": "haze-playout",
                "subject": request_id,
                "data": {
                    "request_id": request_id,
                    "feed_id": request.feed_id,
                    "package_id": request.package_id,
                    "pkg_id": request.package_id,
                    "force": request.force,
                }
            }))
            .await;
        if let Err(err) = publish_result {
            self.pending_products.lock().await.remove(&request_id);
            return Err(err);
        }
        let result = match timeout(PRODUCT_REPLY_TIMEOUT, rx).await {
            Ok(Ok(result)) => result,
            Ok(Err(_)) => {
                self.pending_products.lock().await.remove(&request_id);
                anyhow::bail!("event bridge closed while waiting for product render");
            }
            Err(_) => {
                self.pending_products.lock().await.remove(&request_id);
                anyhow::bail!("product render request timed out");
            }
        };
        match result {
            Ok(product) => Ok(product),
            Err(err) => anyhow::bail!("{err}"),
        }
    }
}

async fn read_loop(
    reader: OwnedReadHalf,
    tx: mpsc::Sender<Value>,
    pending_synth: PendingSynth,
    pending_products: PendingProducts,
) {
    let mut lines = BufReader::new(reader).lines();
    while let Ok(Some(line)) = lines.next_line().await {
        let line = line.trim();
        if line.is_empty() {
            continue;
        }
        let Ok(value) = serde_json::from_str::<Value>(line) else {
            continue;
        };
        if handle_synth_result(&value, &pending_synth).await
            || handle_product_result(&value, &pending_products).await
        {
            continue;
        }
        if tx.send(value).await.is_err() {
            break;
        }
    }
    fail_pending(&pending_synth, "event bridge closed while waiting for TTS").await;
    fail_pending(
        &pending_products,
        "event bridge closed while waiting for product render",
    )
    .await;
}

async fn handle_synth_result(value: &Value, pending: &PendingSynth) -> bool {
    let msg_type = string_at(value, "type");
    if msg_type != "tts.synthesized" && msg_type != "tts.failed" {
        return false;
    }
    let data = data(value);
    let job_id = first_text(value, data, &["job_id", "subject"]);
    if job_id.is_empty() {
        return true;
    }
    let Some(ch) = pending.lock().await.remove(job_id) else {
        return true;
    };
    let result = if msg_type == "tts.failed" {
        Err(format!(
            "TTS failed for {job_id}: {}",
            string_at(data, "error")
        ))
    } else {
        Ok(string_at(data, "output_path").to_string())
    };
    let _ = ch.send(result);
    true
}

async fn handle_product_result(value: &Value, pending: &PendingProducts) -> bool {
    let msg_type = string_at(value, "type");
    if msg_type != "product.rendered" && msg_type != "product.render.failed" {
        return false;
    }
    let data = data(value);
    let request_id = first_text(value, data, &["request_id", "subject"]);
    if request_id.is_empty() {
        return true;
    }
    let Some(ch) = pending.lock().await.remove(request_id) else {
        return true;
    };
    let result = if msg_type == "product.render.failed" {
        Err(format!(
            "product render failed for {request_id}: {}",
            string_at(data, "error")
        ))
    } else {
        serde_json::from_value::<RenderedProduct>(
            data.get("product").cloned().unwrap_or(Value::Null),
        )
        .map_err(|err| err.to_string())
    };
    let _ = ch.send(result);
    true
}

async fn fail_pending<T>(pending: &PendingMap<T>, message: &str) {
    let mut pending = pending.lock().await;
    for (_, ch) in pending.drain() {
        let _ = ch.send(Err(message.to_string()));
    }
}

pub(crate) fn string_at<'a>(value: &'a Value, key: &str) -> &'a str {
    value.get(key).and_then(Value::as_str).unwrap_or("").trim()
}

pub(crate) fn data(value: &Value) -> &Value {
    value.get("data").unwrap_or(&Value::Null)
}

pub(crate) fn first_text<'a>(message: &'a Value, data: &'a Value, keys: &[&str]) -> &'a str {
    for key in keys {
        let from_data = string_at(data, key);
        if !from_data.is_empty() {
            return from_data;
        }
        let from_message = string_at(message, key);
        if !from_message.is_empty() {
            return from_message;
        }
    }
    ""
}

fn fallback_text(value: &str, fallback: &str) -> String {
    let value = value.trim();
    if value.is_empty() {
        fallback.to_string()
    } else {
        value.to_string()
    }
}

File: services/go/internal/playlist/bridge.go
package playlist

import (
	"bufio"
	"context"
	"encoding/json"
	"fmt"
	"log"
	"net"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"
)

type bridgeClient struct {
	conn            net.Conn
	events          chan map[string]any
	pendingProducts map[string]chan productResult
	pendingSynth    map[string]chan synthResult
	mu              sync.Mutex
}

type productResult struct {
	Product renderedProduct
	Err     error
}

type synthResult struct {
	Path string
	Err  error
}

type renderedProduct struct {
	ID        string            `json:"id"`
	FeedID    string            `json:"feed_id"`
	PackageID string            `json:"package_id"`
	Title     string            `json:"title"`
	Text      string            `json:"text"`
	ReaderID  string            `json:"reader_id"`
	Language  string            `json:"language"`
	Metadata  map[string]string `json:"metadata"`
}

type synthJob struct {
	ID         string
	Text       string
	ReaderID   string
	Language   string
	Timezone   string
	OutputPath string
}

func connectBridge(ctx context.Context, addr string) (*bridgeClient, error) {
	if strings.TrimSpace(addr) == "" {
		return nil, fmt.Errorf("missing host event bridge address")
	}
	dialer := net.Dialer{Timeout: 3 * time.Second}
	conn, err := dialer.DialContext(ctx, "tcp", addr)
	if err != nil {
		return nil, err
	}
	client := &bridgeClient{
		conn:            conn,
		events:          make(chan map[string]any, 8192),
		pendingProducts: map[string]chan productResult{},
		pendingSynth:    map[string]chan synthResult{},
	}
	go client.readLoop()
	return client, nil
}

func (c *bridgeClient) Close() error {
	if c == nil || c.conn == nil {
		return nil
	}
	return c.conn.Close()
}

func (c *bridgeClient) Events() <-chan map[string]any {
	return c.events
}

func (c *bridgeClient) Publish(message map[string]any) error {
	c.mu.Lock()
	defer c.mu.Unlock()
	if _, ok := message["timestamp"]; !ok {
		message["timestamp"] = time.Now().UTC()
	}
	return json.NewEncoder(c.conn).Encode(message)
}

func (c *bridgeClient) RenderProduct(ctx context.Context, requestID string, feedID string, packageID string, language string) (renderedProduct, error) {
	ch := make(chan productResult, 1)
	c.mu.Lock()
	c.pendingProducts[requestID] = ch
	c.mu.Unlock()
	defer func() {
		c.mu.Lock()
		delete(c.pendingProducts, requestID)
		c.mu.Unlock()
	}()
	if err := c.Publish(map[string]any{
		"type":    "product.render.request",
		"source":  serviceID,
		"subject": requestID,
		"data": map[string]any{
			"request_id": requestID,
			"feed_id":    feedID,
			"package_id": packageID,
			"pkg_id":     packageID,
			"language":   language,
		},
	}); err != nil {
		return renderedProduct{}, err
	}
	select {
	case <-ctx.Done():
		return renderedProduct{}, ctx.Err()
	case result := <-ch:
		return result.Product, result.Err
	}
}

func (c *bridgeClient) Synthesize(ctx context.Context, job synthJob) (string, error) {
	if strings.TrimSpace(job.ID) == "" {
		job.ID = fmt.Sprintf("tts-%d", time.Now().UnixNano())
	}
	ch := make(chan synthResult, 1)
	c.mu.Lock()
	c.pendingSynth[job.ID] = ch
	c.mu.Unlock()
	defer func() {
		c.mu.Lock()
		delete(c.pendingSynth, job.ID)
		c.mu.Unlock()
	}()
	if job.OutputPath != "" {
		if err := os.MkdirAll(filepath.Dir(job.OutputPath), 0o755); err != nil {
			return "", err
		}
	}
	if err := c.Publish(map[string]any{
		"type":    "tts.synthesize",
		"source":  serviceID,
		"subject": job.ID,
		"data": map[string]any{
			"job_id":      job.ID,
			"text":        job.Text,
			"reader_id":   job.ReaderID,
			"language":    job.Language,
			"timezone":    job.Timezone,
			"output_path": job.OutputPath,
			"priority":    "high",
		},
	}); err != nil {
		return "", err
	}
	select {
	case <-ctx.Done():
		return "", ctx.Err()
	case result := <-ch:
		return result.Path, result.Err
	}
}

func (c *bridgeClient) readLoop() {
	defer close(c.events)
	scanner := bufio.NewScanner(c.conn)
	scanner.Buffer(make([]byte, 64*1024), 4*1024*1024)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}
		var message map[string]any
		if err := json.Unmarshal([]byte(line), &message); err != nil {
			continue
		}
		if c.handleProductResult(message) || c.handleSynthResult(message) {
			continue
		}
		c.enqueueEvent(message)
	}
	c.mu.Lock()
	for id, ch := range c.pendingProducts {
		ch <- productResult{Err: fmt.Errorf("event bridge closed while waiting for %s", id)}
		delete(c.pendingProducts, id)
	}
	for id, ch := range c.pendingSynth {
		ch <- synthResult{Err: fmt.Errorf("event bridge closed while waiting for %s", id)}
		delete(c.pendingSynth, id)
	}
	c.mu.Unlock()
}

func (c *bridgeClient) enqueueEvent(message map[string]any) {
	select {
	case c.events <- message:
		return
	default:
	}
	msgType := stringAt(message, "type")
	if !playlistLifecycleEvent(msgType) {
		log.Printf("playlist bridge event buffer full; dropped %s", msgType)
		return
	}
	select {
	case <-c.events:
	default:
	}
	select {
	case c.events <- message:
	default:
		log.Printf("playlist bridge event buffer full; dropped %s", msgType)
	}
}

func playlistLifecycleEvent(msgType string) bool {
	switch strings.TrimSpace(msgType) {
	case "playout.accepted", "playout.started", "playout.interrupted", "playout.completed",
		"alert.playout.started", "alert.playout.completed", "service.ready", "system.shutdown":
		return true
	default:
		return false
	}
}

func (c *bridgeClient) handleProductResult(message map[string]any) bool {
	msgType := stringAt(message, "type")
	if msgType != "product.rendered" && msgType != "product.render.failed" {
		return false
	}
	data := mapAt(message, "data")
	requestID := firstText(message, data, "request_id", "subject")
	if requestID == "" {
		return true
	}
	c.mu.Lock()
	ch := c.pendingProducts[requestID]
	c.mu.Unlock()
	if ch == nil {
		return true
	}
	if msgType == "product.render.failed" {
		ch <- productResult{Err: fmt.Errorf("product render failed: %s", stringAt(data, "error"))}
		return true
	}
	raw, err := json.Marshal(data["product"])
	if err != nil {
		ch <- productResult{Err: err}
		return true
	}
	var product renderedProduct
	if err := json.Unmarshal(raw, &product); err != nil {
		ch <- productResult{Err: err}
		return true
	}
	ch <- productResult{Product: product}
	return true
}

func (c *bridgeClient) handleSynthResult(message map[string]any) bool {
	msgType := stringAt(message, "type")
	if msgType != "tts.synthesized" && msgType != "tts.failed" {
		return false
	}
	data := mapAt(message, "data")
	jobID := firstText(message, data, "job_id", "subject")
	if jobID == "" {
		return true
	}
	c.mu.Lock()
	ch := c.pendingSynth[jobID]
	c.mu.Unlock()
	if ch == nil {
		return true
	}
	if msgType == "tts.failed" {
		ch <- synthResult{Err: fmt.Errorf("TTS failed: %s", stringAt(data, "error"))}
		return true
	}
	ch <- synthResult{Path: stringAt(data, "output_path")}
	return true
}

func stringAt(source map[string]any, key string) string {
	if source == nil {
		return ""
	}
	switch value := source[key].(type) {
	case string:
		return strings.TrimSpace(value)
	default:
		return ""
	}
}

func mapAt(source map[string]any, key string) map[string]any {
	if source == nil {
		return nil
	}
	if value, ok := source[key].(map[string]any); ok {
		return value
	}
	return nil
}

func firstText(message map[string]any, data map[string]any, keys ...string) string {
	for _, key := range keys {
		if value := stringAt(message, key); value != "" {
			return value
		}
		if value := stringAt(data, key); value != "" {
			return value
		}
	}
	return ""
}

func firstValue(message map[string]any, data map[string]any, keys ...string) any {
	for _, key := range keys {
		if message != nil {
			if value, ok := message[key]; ok && value != nil {
				return value
			}
		}
		if data != nil {
			if value, ok := data[key]; ok && value != nil {
				return value
			}
		}
	}
	return nil
}

func boolAny(value any) bool {
	switch typed := value.(type) {
	case bool:
		return typed
	case string:
		switch strings.ToLower(strings.TrimSpace(typed)) {
		case "1", "true", "yes", "y", "on", "broadcast immediate":
			return true
		}
	}
	return false
}

func stringListAny(value any) []string {
	switch typed := value.(type) {
	case []any:
		out := make([]string, 0, len(typed))
		for _, item := range typed {
			text := strings.TrimSpace(fmt.Sprint(item))
			if text != "" {
				out = append(out, text)
			}
		}
		return out
	case []string:
		out := make([]string, 0, len(typed))
		for _, item := range typed {
			text := strings.TrimSpace(item)
			if text != "" {
				out = append(out, text)
			}
		}
		return out
	case string:
		if text := strings.TrimSpace(typed); text != "" {
			return []string{text}
		}
	}
	return nil
}
